### PR TITLE
#30 Add JSON case reports

### PR DESCRIPTION
**Pull request by WGierke**
This PR is based on and blocked by #93 (because this already adds a proper creation of candidates). To see the raw changes this PR wants to add, please see [here](https://github.com/WGierke/concept-to-clinic/compare/36_accept_reject_candidates...WGierke:30_case_json_report).
Now, a JSON report can be generated for a case that includes the patient id, candidates, nodules, etc.

## Description
I fixed the creation of nodules, added the JSON report and wrote a test for that.

## Reference to official issue
This addresses #30.

## Motivation and Context
This should summarize the data of a case and help debugging.

## How Has This Been Tested?
I wrote a test which adds a case with image series, candidates and nodules and which tests for a proper report.

## Screenshots (if appropriate):
screenshot from 2017-09-06 15-43-02

## CLA
- [X] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

**Comment by lamby** [comment]
@WGierke Still conflicting; please rebase :)

**Comment by lamby** [comment]
@WGierke Haha. Oh dear. And I just tried to merge them using github's new UI which didn't seem to do anything :p

File: interface/backend/api/serializers.py [hunk 1]
```diff
@@ -54,3 +54,11 @@ class Meta:
         read_only_fields = ('created',)
 
     centroid = ImageLocationSerializer()
+
+    def create(self, validated_data):
+        case_data = validated_data.pop('case')
```
Comment (lamby) [inline interface/backend/api/serializers.py]: Why `pop` the data? I mean, versus just indexing it with `validated_data['case']` (ie. please add a comment to the code itself if there is a reason; this is non-intuitive)

File: interface/backend/api/views.py [hunk 2]
```diff
@@ -55,3 +61,38 @@ def candidate_mark(request, candidate_id):
...
+    case_dict['candidates'] = candidates
+
+    nodules = [model_to_dict(nodule) for nodule in Nodule.objects.filter(case_id=case_id)]
+    for nodule in nodules:
+        nodule['centroid'] = model_to_dict(ImageLocation.objects.get(pk=nodule['centroid']))
+    case_dict['nodules'] = nodules
+    return case_dict
+
+
+def json_serial(obj):
```
Comment (lamby) [inline interface/backend/api/views.py]: Looks like something we want in a globalish `utils` file instead of here. :)

File: interface/backend/api/views.py [hunk 3]
```diff
@@ -55,3 +61,38 @@ def candidate_mark(request, candidate_id):
...
+
+    if isinstance(obj, (datetime, date)):
+        return obj.isoformat()
+    raise TypeError("Type %s not serializable" % type(obj))
+
+
+def case_report(request, case_id):
+    case_dict = generate_case_dict(case_id)
+    case_pretty_json = json.dumps(case_dict, indent=4, sort_keys=True, default=json_serial)
+    return HttpResponse(case_pretty_json, content_type="application/json")
```
Comment (lamby) [inline interface/backend/api/views.py]: Hm, do we not have a `JsonResponse` ?

Comment (WGierke) [reply interface/backend/api/views.py]: I'd also prefer pure JSON put the issue states `For now, this can be a simple view that takes a Case data, generates the corresponding dict structure and pretty-prints using the pprint module within a HTML page.` Doesn't that mean the rendered page should be a HTML page including the pretty printed JSON?

File: interface/backend/api/serializers.py (hunk 1 above)
Comment (WGierke) [reply interface/backend/api/serializers.py]: :man_facepalming: Yep, `pop`ing definitely makes no sense here :)

File: interface/backend/api/views.py (hunk 3 above)
Comment (WGierke) [reply interface/backend/api/views.py]: In fact, sending `application/json` as content type makes it possible to use browser add-ons that make JSON explorable and even prettier:
screenshot from 2017-09-10 19-27-55
Without such add-ons, the JSON is still printed in a nice way due to the HttpResponse.
screenshot from 2017-09-10 19-28-09

Comment (lamby) [reply interface/backend/api/views.py]: But surely JsonResponse sends exactly that? (Or we should have our own version?)

Comment (WGierke) [reply interface/backend/api/views.py]: When I use JsonResponse and try to pretty-print the JSON using `json.dumps(case_dict, indent=4, sort_keys=True..`, I end up with 
screenshot from 2017-09-11 08-51-21
which is obviously neither pretty nor valid JSON.
Without using `json.dumps` and without using a browser add-on for pretty JSON formatting, you end up with
screenshot from 2017-09-11 08-49-56
which is also not pretty.
So I'd clearly plead for the current HttpResponse and json.dumps implementation.

Comment (isms) [reply interface/backend/api/views.py]: @WGierke (cc @lamby) Seems to me neither `HttpResponse` nor `JsonResponse` is the idiomatic approach here.

Let's do this the DRF way, yeah? Please have a look at the DRF documentation, this is a solved problem: http://www.django-rest-framework.org/tutorial/2-requests-and-responses/

File: interface/backend/api/views.py [hunk 4]
```diff
@@ -55,3 +60,30 @@ def candidate_mark(request, candidate_id):
 
 def candidate_dismiss(request, candidate_id):
     return JsonResponse({'response': "Candidate {} was dismissed".format(candidate_id)})
+
+
+def generate_case_dict(case_id):
+    try:
+        case_dict = model_to_dict(Case.objects.get(pk=case_id))
+    except:
```
Comment (lamby) [inline interface/backend/api/views.py]: ``except`` is too broad an exception, surely?

File: interface/utils.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from datetime import date, datetime
+
+
+def json_serial(obj):
+    """JSON serializer for objects not serializable by default json code"""
+
+    if isinstance(obj, (datetime, date)):
```
Comment (lamby) [inline interface/utils.py]: Aaaaaaactually, isn't there one of these in Django?

File: interface/backend/api/views.py (hunk 3 above)
Comment (WGierke) [reply interface/backend/api/views.py]: Thanks @isms I wasn't aware of that idiom.
Now one can add a suffix that states what content type should be returned:
screenshot from 2017-09-12 09-53-32
screenshot from 2017-09-12 09-53-47

File: interface/utils.py (hunk 5 above)
Comment (WGierke) [reply interface/utils.py]: Yes, I found out that `DjangoJSONEncoder` also works here :)

**Comment by isms** [comment]
To be clear about what the DRF best practice is, models should be serialized with a ModelSerializer, not a custom function. Then there are lots of ways to serve that. You also don't want to manually specify the content type in general - let DRF negotiate that so it can provide the dev interface.

File: docs/apidoc_interface/backend.rst [hunk 6]
```diff
@@ -9,7 +9,6 @@ Subpackages
     backend.api
     backend.cases
     backend.images
-    backend.static
```
Comment (lamby) [inline docs/apidoc_interface/backend.rst]: Not sure why this is in here? :)

File: interface/backend/api/urls.py [hunk 7]
```diff
@@ -26,3 +28,6 @@
     url(r'^candidates/(?P<candidate_id>\d+)/dismiss$', candidate_dismiss, name='candidate-dismiss'),
     url(r'^candidates/(?P<candidate_id>\d+)/mark$', candidate_mark, name='candidate-mark'),
 ]
+
+# Support different suffixes
+urlpatterns.extend(format_suffix_patterns([url(r'^cases/(?P<case_id>\d+)/report$', case_report, name='case-report')]))
```
Comment (lamby) [inline interface/backend/api/urls.py]: FYI the typical pattern is to use ``+=``

File: interface/backend/api/views.py [hunk 8]
```diff
@@ -89,3 +97,23 @@ def candidate_mark(request, candidate_id):
...
+    def render(self, data, media_type=None, renderer_context=None):
+        return "<pre>{}</pre>".format(json.dumps(data, indent=4, sort_keys=True, cls=DjangoJSONEncoder))
+
+
+@api_view(['GET'])
+# Render .json and .html requests
+@renderer_classes((JSONRenderer, JsonHtmlRenderer))
+def case_report(request, case_id, format=None):
+    try:
+        case = Case.objects.get(pk=case_id)
```
Comment (lamby) [inline interface/backend/api/views.py]: I think you want ``get_object_or_404`` here

File: interface/backend/api/serializers.py [hunk 9]
```diff
@@ -54,3 +54,11 @@ class Meta:
         read_only_fields = ('created',)
 
     centroid = ImageLocationSerializer()
+
+    def create(self, validated_data):
```
Comment (lamby) [inline interface/backend/api/serializers.py]: This seems... quite long-winded. Why not just:

```
def create(self, validated_data):
    return Nodule.objects.create(
        case=validated_data['case'],
        candidate=validated_data['candidate'],
        centroid=ImageLocation.objects.create(**validated_data['centroid']),
    )

It's currently quite hard to follow with all the variable assignments to sliiiiiightly different names if you know what I mean?

Comment (WGierke) [reply interface/backend/api/serializers.py]: Nice. Short and exactly what I tried to achieve ... Thanks!

**Comment by WGierke** [comment]
@lamby Thanks again for suggesting some really nice and clean shortcuts. I implemented them and squashed the commits.

**Comment by lamby** [comment]
Thanks!
